sympy/ntheory/generate.py

```diff
@@ -745,14 +747,37 @@ def primerange(a, b=None):
         a, b = 2, a
     if a >= b:
         return
-    # if we already have the range, return it
-    if b <= sieve._list[-1]:
+    # If we already have the range, return it.
+    largest_known_prime = sieve._list[-1]
+    if b <= largest_known_prime:
         yield from sieve.primerange(a, b)
         return
+    # If we know some of it, return it.
+    if a <= largest_known_prime:
+        yield from sieve.primerange(a, largest_known_prime + 1)
+        a = largest_known_prime + 1
+    if a % 2:
+        a -= 1
+    # Array length to be allocated. The larger the length,
```

Comment:
It seems like this must be duplicating something somewhere. Since this routine didn't formerly handle growing the sieve, why does it need to do so now? Should this growth mechanism be abstracted to a method for the sieve itself rather than be part of the `primerange` function?

Reply:
That makes sense to me. I am in favor of that change.